src/DSA.jl

```diff
@@ -1,174 +1,313 @@
 # This file is a part of JuliaFEM.
 # License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
 
-@with_kw mutable struct DSADriverState <: AbstractMaterialState
-    time::Float64 = zero(Float64)
-    strain::SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module DSAModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericDSA, GenericDSADriverState, GenericDSAParameterState, GenericDSAVariableState
+
+# specialization for Float64
+export DSA, DSADriverState, DSAParameterState, DSAVariableState
+
+@with_kw mutable struct GenericDSADriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+end
+
+# TODO: complete this docstring
+"""Parameter state for DSA (dynamic strain aging) material.
+
+This is similar to the Chaboche model, but with additional static recovery terms.
+
+`E`: Young's modulus
+`nu`: Poisson's ratio
+`R0`: initial yield strength
+`Kn`: plasticity multiplier divisor (drag stress)
+`nn`: plasticity multiplier exponent
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q`: shift parameter for yield strength evolution
+`b`: multiplier for yield strength evolution
+`w`: ???
+`P1`, `P2`: ???
+`m`: ???
+`m1`, `m2`: ???
+`M1`, `M2`: ???
+`ba`: ???
+`xi`: ???
```

Comment:
A dimensionless parameter that controls the magnitude of the Marquis effect from the aging stress (whether the increased hardening due to aging shows as increased relaxation or not). Support [0,1]. With the value of 0, the aging stress contributes solely to the size of the yield surface `R` (isotropic hardening). With the value of 1, the aging stress contributes solely to the viscoplastic drag stress `K`.